src/metabase/db/liquibase.clj

```diff
@@ -197,31 +199,39 @@
   "Check and make sure the database isn't locked. If it is, sleep for 2 seconds and then retry several times. There's a
   chance the lock will end up clearing up so we can run migrations normally."
   [^Liquibase liquibase]
-  (u/auto-retry 5
-    (when (migration-lock-exists? liquibase)
-      (Thread/sleep 2000)
-      (throw
-       (LockException.
-        (str
-         (trs "Database has migration lock; cannot run migrations.")
-         " "
-         (trs "You can force-release these locks by running `java -jar metabase.jar migrate release-locks`.")))))))
+  (if (migration-lock-exists? liquibase)
+    (do
+     (log/info "Database has migration lock. Waiting for lock to be cleared...")
+     (Thread/sleep 2000)
+     (u/auto-retry 4
+       (when (migration-lock-exists? liquibase)
+         (Thread/sleep 2000)
+         (throw
+          (LockException.
+           (str
+            (trs "Database has migration lock; cannot run migrations.")
+            " "
+            (trs "You can force-release these locks by running `java -jar metabase.jar migrate release-locks`.")))))))
+    (log/info "No migration lock found.")))
 
 (defn migrate-up-if-needed!
   "Run any unrun `liquibase` migrations, if needed."
-  [^Liquibase liquibase]
+  [^Liquibase liquibase ^DataSource data-source]
   (log/info (trs "Checking if Database has unrun migrations..."))
-  (if (seq (unrun-migrations liquibase))
+  (if (seq (unrun-migrations data-source))
     (do
-     (log/info (trs "Database has unrun migrations. Waiting for migration lock to be cleared..."))
+     (log/info (trs "Database has unrun migrations. Checking if migraton lock is cleared..."))
      (wait-for-migration-lock-to-be-cleared liquibase)
     ;; while we were waiting for the lock, it was possible that another instance finished the migration(s), so make
     ;; sure something still needs to be done...
-     (let [unrun-migrations-count (count (unrun-migrations liquibase))]
+     (let [to-run-migrations      (unrun-migrations data-source)
+           unrun-migrations-count (count to-run-migrations)]
        (if (pos? unrun-migrations-count)
          (let [^Contexts contexts nil
                start-time         (System/currentTimeMillis)]
-           (log/info (trs "Migration lock is cleared. Running {0} migrations ..." unrun-migrations-count))
+           (log/info (trs "Running {0} migrations ..." unrun-migrations-count))
+           (doseq [change to-run-migrations]
```

Comment:
I'll remove this before merging, leaving it here for debugging purposes

Reply:
What about leaving it as a `trace` statement? Do we have a good mechanism for dynamically toggling log levels per namespace?

Reply:
yea trace is a good call.

> Do we have a good mechanism for dynamically toggling log levels per namespace?

you can use `metabase.test/set-ns-log-level!` in the REPL, or update `resources/log4j2.xml` if you use a jar